File: java/v1/src/main/java/com/api/v1/utils/car/CarFinderUtil.java
package com.api.v1.utils.car;


import com.api.v1.annotations.car.VIN;
import com.api.v1.domain.car.Car;
import com.api.v1.domain.car.CarRepository;
import com.api.v1.exceptions.car.CarNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
public class CarFinderUtil {

    @Autowired
    private CarRepository repository;

    public Mono<Car> find(@VIN String vin) {
        return repository
                .findAll()
                .filter(e -> e.getVin().equals(vin))
                .singleOrEmpty()
                .switchIfEmpty(Mono.error(new CarNotFoundException(vin)));
    }

}

File: kotlin/v2/src/main/kotlin/com/api/v2/car/utils/CarFinderUtil.kt
package com.api.v2.car.utils

import com.api.v2.car.exceptions.CarNotFoundException
import com.api.v2.car.domain.Car
import com.api.v2.car.domain.CarRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class CarFinderUtil {

    @Autowired
    private lateinit var carRepository: CarRepository

    suspend fun find(vin: String): Car {
        return withContext(Dispatchers.IO) {
            carRepository
                .findAll()
                .filter { e -> e.vin ==  vin }
                .firstOrNull() ?: throw CarNotFoundException(vin)
        }

    }

}